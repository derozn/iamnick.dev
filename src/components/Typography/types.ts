export type ComponentToVariantMapping = {
  h1: string;
  h2: string;
  h3: string;
  p: string;
  [key: string]: string;
};

export type Display = 'block' | 'inline';
export type Align = 'left' | 'center' | 'right' | 'justify';
export type Component = 'h1' | 'h2' | 'h3' | 'p';
export type Color = 'primary' | 'secondary' | 'accent';

export type Props = {
  align?: Align;
  color?: Color;
  variant?: string;
  component?: Component;
  display?: Display;
};

export type TextProps = {
  variant?: string;
  textColor?: Color;
  textAlign?: Align;
  displayAs?: Display;
};
