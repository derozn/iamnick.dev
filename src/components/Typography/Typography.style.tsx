/* eslint-disable complexity */
import styled, { css } from 'styled-components';
import { TextProps } from './types';

export const Text = styled.span<TextProps>(
  (props) => css`
  margin: 0;
  position: relative;

  ${props.displayAs === 'block' && `display: block;`}
  ${props.displayAs === 'inline' && `display: inline-block;`}

  ${props.variant === 'h1' && props.theme.typography.h1}
  ${props.variant === 'h2' && props.theme.typography.h2}
  ${props.variant === 'h3' && props.theme.typography.h3}
  ${props.variant === 'body' && props.theme.typography.body}

  ${props.textColor === 'primary' && `color: ${props.theme.palette.text.primary};`}
  ${props.textColor === 'secondary' && `color: ${props.theme.palette.text.secondary};`}
  ${props.textColor === 'accent' && `color: ${props.theme.palette.text.accent};`}
  
  ${props.textAlign && `text-align: ${props.textAlign};`}
`,
);
