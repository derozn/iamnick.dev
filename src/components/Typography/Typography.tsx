import React, { FunctionComponent } from 'react';

import { Props, ComponentToVariantMapping } from './types';
import { Text } from './Typography.style';

const componentToVariantMapping: ComponentToVariantMapping = {
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  p: 'body',
};

const Typography: FunctionComponent<Props> = (props) => {
  const { children, variant, component, color, align, display, ...rest } = props;

  const variantClass = variant || (component && componentToVariantMapping[component]);

  return (
    <Text
      as={component}
      variant={variantClass}
      textColor={color}
      textAlign={align}
      displayAs={display}
      {...rest}
    >
      {children}
    </Text>
  );
};

export default Typography;
