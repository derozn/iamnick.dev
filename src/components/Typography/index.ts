export { default } from './Typography';
export * from './types';
