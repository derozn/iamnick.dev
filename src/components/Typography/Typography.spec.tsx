import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { theme } from '#styles';

import Typography from './Typography';

describe('components/Typography', () => {
  it('renders as span', () => {
    const { container } = render(<Typography />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders children', () => {
    const { getByText } = render(<Typography>hello i am nick</Typography>);

    expect(getByText('hello i am nick')).toBeTruthy();
  });

  it('renders as h1', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Typography component="h1" />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders as h2', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Typography component="h2" />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders as h3', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Typography component="h3" />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders as p', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Typography component="p" />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders as h1 styles when variant is h1', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Typography component="h2" variant="h1" />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with primary text color when color is primary', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Typography color="primary" />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with secondary text color when color is secondary', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Typography color="secondary" />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with textAlign', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Typography align="center" />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with display', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Typography display="inline" />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
