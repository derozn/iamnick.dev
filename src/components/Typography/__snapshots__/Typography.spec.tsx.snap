// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Typography renders as h1 1`] = `
.c0 {
  margin: 0;
  position: relative;
  font-family: 'Montserrat',sans-serif;
  font-weight: 600;
  font-size: 4.8rem;
  line-height: 1.167;
  -webkit-letter-spacing: -0.01562em;
  -moz-letter-spacing: -0.01562em;
  -ms-letter-spacing: -0.01562em;
  letter-spacing: -0.01562em;
}

@media (min-width:600px) {
  .c0 {
    font-size: 6rem;
  }
}

@media (min-width:1280px) {
  .c0 {
    font-size: 7.2rem;
  }
}

<h1
  class="c0"
/>
`;

exports[`components/Typography renders as h1 styles when variant is h1 1`] = `
.c0 {
  margin: 0;
  position: relative;
  font-family: 'Montserrat',sans-serif;
  font-weight: 600;
  font-size: 4.8rem;
  line-height: 1.167;
  -webkit-letter-spacing: -0.01562em;
  -moz-letter-spacing: -0.01562em;
  -ms-letter-spacing: -0.01562em;
  letter-spacing: -0.01562em;
}

@media (min-width:600px) {
  .c0 {
    font-size: 6rem;
  }
}

@media (min-width:1280px) {
  .c0 {
    font-size: 7.2rem;
  }
}

<h2
  class="c0"
/>
`;

exports[`components/Typography renders as h2 1`] = `
.c0 {
  margin: 0;
  position: relative;
  font-family: 'Montserrat',sans-serif;
  font-weight: 400;
  font-size: 3.75rem;
  line-height: 1.2;
  -webkit-letter-spacing: -0.00833em;
  -moz-letter-spacing: -0.00833em;
  -ms-letter-spacing: -0.00833em;
  letter-spacing: -0.00833em;
}

@media (min-width:600px) {
  .c0 {
    font-size: 4.8rem;
  }
}

@media (min-width:1280px) {
  .c0 {
    font-size: 6rem;
  }
}

<h2
  class="c0"
/>
`;

exports[`components/Typography renders as h3 1`] = `
.c0 {
  margin: 0;
  position: relative;
  font-family: 'Montserrat',sans-serif;
  font-weight: 400;
  font-size: 2.8rem;
  line-height: 1.167;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
}

@media (min-width:1280px) {
  .c0 {
    font-size: 3rem;
  }
}

<h3
  class="c0"
/>
`;

exports[`components/Typography renders as p 1`] = `
.c0 {
  margin: 0;
  position: relative;
  font-family: 'Open Sans',arial,sans-serif;
  font-weight: 400;
  font-size: 1.6rem;
  line-height: 1.5;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
}

@media (min-width:960px) {
  .c0 {
    font-size: 1.8rem;
  }
}

<p
  class="c0"
/>
`;

exports[`components/Typography renders as span 1`] = `
.c0 {
  margin: 0;
  position: relative;
}

<span
  class="c0"
/>
`;

exports[`components/Typography renders with display 1`] = `
.c0 {
  margin: 0;
  position: relative;
  display: inline-block;
}

<span
  class="c0"
/>
`;

exports[`components/Typography renders with primary text color when color is primary 1`] = `
.c0 {
  margin: 0;
  position: relative;
  color: #d6d6d6;
}

<span
  class="c0"
/>
`;

exports[`components/Typography renders with secondary text color when color is secondary 1`] = `
.c0 {
  margin: 0;
  position: relative;
  color: #000000;
}

<span
  class="c0"
/>
`;

exports[`components/Typography renders with textAlign 1`] = `
.c0 {
  margin: 0;
  position: relative;
  text-align: center;
}

<span
  class="c0"
/>
`;
